### PR TITLE
Add retry argument builder unit tests

### LINKED ISSUE #10563: Retry orchestrator argument-cleanup logic (RetryArgumentsBuilder/RetryOrchestratorHelper) has zero unit test coverage
### 🎯 Repository Quality Improvement Report — Retry Orchestrator Argument-Building Unit Test Gap

**Analysis Date**: 2026-08-11
**Focus Area**: `retry-orchestrator-argument-builder-coverage-gap` (custom)
**Strategy Type**: Custom

### Executive Summary

`Microsoft.Testing.Extensions.Retry` (the `--retry-failed-tests` MTP extension) relies on two internal static helper classes — `RetryArgumentsBuilder` and `RetryOrchestratorHelper` — to compute which original command-line arguments must be stripped before relaunching a failed test host, and to rebuild the per-attempt argument list (result directory, pipe name, `--filter-uid` list or response-file fallback). This logic is genuinely tricky: it juggles index-based cleanup ordering, quoted-UID detection, a 30,000-character response-file threshold, and stripping of `--minimum-expected-tests` on retry (a behavior added specifically to fix issue #5639).

Despite that complexity, `test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryTests.cs` (300 lines, 16 test methods) contains **no references at all** to `RetryArgumentsBuilder`, `ComputeIndicesToCleanup`, `BuildAttemptArgumentsAsync`, `RetryOrchestratorHelper.RemoveOption`, `GetOptionArgumentIndex`, `FormatDelay`, or `FormatDuration`. A full-repo search confirms zero test files reference these APIs directly. The only indirect coverage is via slow, process-spawning acceptance tests in `RetryFailedTestsTests.cs` (966 lines) that exercise end-to-end retry scenarios but cannot cheaply cover boundary conditions like the exact response-file length threshold or UIDs containing embedded double quotes.

This is a coverage gap with real risk: a regression in `RemoveOption`'s handling of `--option=value` vs. `--option:value` forms, or in the response-file quoting logic for UIDs containing `"`, would only be caught by a slow acceptance test (if at all), and such regressions could silently corrupt the argument list passed to the retried test host — causing wrong tests to run or the process to crash with a cryptic parser error.

<details>
<summary><b>Full Analysis Report</b></summary>

### Current State Assessment

**Metrics Collected:**
| Metric | Value | Status |
|--------|-------|--------|
| Unit test references to `RetryArgumentsBuilder` | 0 | ❌ |
| Unit test references to `RetryOrchestratorHelper.RemoveOption`/`GetOptionArgumentIndex` | 0 | ❌ |
| Unit test references to `FormatDelay`/`FormatDuration` (retry package) | 0 | ❌ (sibling `AzureDevOpsSlowTestHistoryTests.cs` tests an analogous `FormatDuration` in a different extension, showing the pattern is known and cheap to apply) |
| Existing `RetryTests.cs` unit tests | 300 lines / 16 methods, none touching argument construction | ⚠️ |
| Acceptance-level retry test coverage | `RetryFailedTestsTests.cs` (966 lines), `RetryWithMaxFailedTestsTests.cs` (223 lines) | ✅ (good for end-to-end, but slow/expensive for edge cases) |
| `RetryThresholdPolicy` (78 lines, threshold math) unit coverage | 0 direct references found | ❌ |

#### Strengths
- Extensive acceptance/integration test coverage for retry end-to-end scenarios, including the `--minimum-expected-tests` stripping fix for #5639.
- The helper classes are already well-isolated as pure, stateless static methods with clear doc comments — ideal candidates for fast unit tests with no process-spawning overhead.
- A sibling extension (`AzureDevOpsSlowTestThresholds.FormatDuration`) already demonstrates the exact unit-test pattern (`[DataRow]`-driven) that should be replicated here.

#### Areas for Improvement
- **[High]** `RetryArgumentsBuilder.ComputeIndicesToCleanup` and `BuildAttemptArgumentsAsync` — the core "which args survive a retry relaunch" logic — has no fast unit test. Edge cases like UIDs with embedded quotes forcing inline-only mode, and the exact `CommandLineLengthLimit` (30,000 char) boundary triggering the response-file fallback, are only reachable via slow acceptance tests that would need enormous generated UID lists to trip the threshold — likely nobody actually exercises that branch today.
- **[Medium]** `RetryOrchestratorHelper.RemoveOption` supports four argument forms (`--opt`, `-opt`, `--opt=value`, `--opt:value`) plus multi-arity trailing values; none of these branches has a direct unit test.
- **[Medium]** `RetryOrchestratorHelper.GetOptionArgumentIndex`, `FormatDelay`, `FormatDuration` are simple, pure, and fast to test but currently untested.
- **[Low]** `RetryThresholdPolicy.EvaluateAsync` mixes percentage/count math with output-device side effects; the pure math portions (rounding, threshold comparisons) could be extracted or at least covered via a mockable `IOutputDevice`.

</details>

---

#### Task 1: Add unit tests for `RetryOrchestratorHelper.RemoveOption`

**Priority**: High
**Estimated Effort**: Small

Add `[DataRow]`-driven tests to `RetryTests.cs` covering: `--opt value`, `-opt value`, `--opt=value`, `--opt:value`, multi-arity options with several trailing values (e.g. `--filter-uid=1 2 3`), options appearing multiple times, and the case where the option is absent (no-op). Assert the resulting `List<string>` exactly.

---

#### Task 2: Add unit tests for `RetryArgumentsBuilder.ComputeIndicesToCleanup`

**Priority**: High
**Estimated Effort**: Small

Cover: no optional options set (only `--retry-failed-tests N` present), all optional options set (`--retry-failed-tests-max-percentage`, `--retry-failed-tests-max-tests`, `--retry-failed-tests-delay`, `--results-directory`), and the edge case where `--retry-failed-tests-delay` is the last argument with no trailing value (line 58's `if (argIndex + 1 < executableArguments.Length)` guard). Assert the returned index list matches expectations.

---

#### Task 3: Add unit tests for `RetryArgumentsBuilder.BuildAttemptArgumentsAsync` response-file fallback

**Priority**: High
**Estimated Effort**: Medium

Using a mocked `IFileSystem` (pattern already used elsewhere in `RetryTests.cs`), test: (a) a small UID list stays inline with `--filter-uid`; (b) a UID list whose predicted length exceeds `CommandLineLengthLimit` (30,000) is written to an `@retry-filter-uids-{n}.rsp` file instead, and the returned arguments contain the `@`-prefixed path; (c) a UID containing a literal `"` forces inline mode (`useResponseFile` stays `false`) even if length would otherwise exceed the threshold — verifying the `hasUidsWithQuotes` short-circuit at line 124; (d) `--minimum-expected-tests` is stripped from `finalArguments` when `lastListOfFailedId` is non-empty (regression guard for #5639), and passes through unchanged on the first attempt (`lastListOfFailedId is null`).

---

#### Task 4: Add unit tests for `RetryOrchestratorHelper.FormatDelay` and `FormatDuration`

**Priority**: Medium
**Estimated Effort**: Small

Mirror the existing `AzureDevOpsSlowTestHistoryTests.FormatDuration_ProducesCompactStrings` pattern with `[DataRow]` cases: sub-second delays (`500ms`), whole seconds (`1s`), non-whole seconds (`1500ms`), and duration formatting boundaries (`<1s`, `<1m`, `>=1m`) per the doc comments in `RetryOrchestratorHelper.cs` lines 62–79.

---

#### Task 5: Add unit tests for `RetryThresholdPolicy.EvaluateAsync`

**Priority**: Medium
**Estimated Effort**: Medium

Using a mocked `RetryFailedTestsPipeServer` (or its underlying data) and a mocked `IOutputDevice`, cover: percentage threshold not tripped, percentage threshold tripped (verify the rounded percentage in the emitted message), count threshold not tripped, count threshold tripped, and the case where neither option is set (returns `false` without producing output). This closes the last untested file in the retry policy chain.

---

### 📊 Historical Context

<details>
<summary><b>Previous Focus Areas (last 10)</b></summary>

| Date | Focus Area | Type |
|------|------------|------|
| 2026-07-30 | assert-interpolated-string-handler-coverage-gap | Custom |
| 2026-08-04 | mtp-exit-code-documentation-gap | Custom |
| 2026-08-05 | cli-extension-cross-option-validation-gap | Custom |
| 2026-08-07 | html-report-artifact-postprocessor-gap | Custom |
| 2026-08-10 | notimplementedexception-landmines-in-extension-surface | Custom |
| 2026-08-11 | retry-orchestrator-argument-builder-coverage-gap | Custom (this run) |

</details>

---

#### Immediate Actions (This Week)
1. Add unit tests for `RemoveOption` and `ComputeIndicesToCleanup` (Tasks 1–2) — Priority: High
2. Add response-file fallback tests for `BuildAttemptArgumentsAsync` (Task 3) — Priority: High

#### Short-term Actions (This Month)
1. Add `FormatDelay`/`FormatDuration` tests (Task 4) — Priority: Medium
2. Add `RetryThresholdPolicy.EvaluateAsync` tests (Task 5) — Priority: Medium

*Next analysis: 2026-08-12 (or next scheduled run) — Focus area selected based on diversity algorithm*

> 🤖 **Automated content by GitHub Copilot.** Generated by the [Repository Quality Improver](https://github.com/microsoft/testfx/actions/runs/31542200903/agentic_workflow) workflow. · auto · 79.8 AIC · ⌖ 3.6 AIC · ⊞ 15.2K · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Frepository-quality-improver%22&type=issues))
>
<details>
<summary><sub>Add this agentic workflow to your repo</sub></summary>

To install this agentic workflow, run

```
gh aw add githubnext/agentics/workflows/repository-quality-improver.md@main
```
</details>

> - [x] expires  on Aug 13, 2026, 10:31 PM UTC

### PR DESCRIPTION
**Pull request by nohwnd**
Add direct unit tests for retry argument cleanup and failed-test argument rebuilding. This covers short and long option forms, stale filter and `--minimum-expected-tests` removal, the first-attempt path where there are no failed IDs, the response-file fallback, and quoted UIDs that must stay inline.

The over-length scenario is driven by the failed-UID payload rather than by an over-long original argument, so it actually exercises the UID-length accounting in `predictedLength`. Verified by deleting that accumulation from `RetryArgumentsBuilder`: `BuildAttemptArgumentsAsync_OverLengthFailedIds_WritesFailedIdsToResponseFile` fails without it and passes with it restored.

Validated with the full unit test build.

Fix #10563

🤖

### CODE REVIEW

File: test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryArgumentsBuilderTests.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        fileStream.SetupGet(stream => stream.Stream).Returns(memoryStream);
+        fileStream.Setup(stream => stream.Dispose());
+        var fileSystem = new Mock<IFileSystem>(MockBehavior.Strict);
+        fileSystem
+            .Setup(fs => fs.NewFileStream(responseFilePath, FileMode.Create, FileAccess.Write))
+            .Returns(fileStream.Object);
+
+        List<string> actual = await RetryArgumentsBuilder.BuildAttemptArgumentsAsync(
+            fileSystem.Object,
+            [CreateOverLengthArgument()],
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryArgumentsBuilderTests.cs]: You're right, and the test was passing for the wrong reason. A single 40,000-character argument is already over the 30,000 limit by itself, so moving the two UIDs into a response file changed nothing, and the UID-length accounting was effectively untested.

The over-length scenario is now driven by the failed UIDs. The original command line is just `test.dll`, and `CreateOverLengthFailedIds` generates enough realistic UIDs that their own `uid.Length + PerArgumentOverhead` crosses the limit. `AssertOnlyFailedIdsExceedLengthLimit` asserts exactly that shape — original arguments under the limit, UID payload over it — so the fixture cannot drift back without failing.

I verified it with the mutation you implied: with the `foreach (string uid in lastListOfFailedId)` accumulation deleted from `predictedLength`, `BuildAttemptArgumentsAsync_OverLengthFailedIds_WritesFailedIdsToResponseFile` fails on the `@...retry-filter-uids-2.rsp` argument being absent. With the line restored, all 33 tests pass again.

The quoted-UID test gets the same fixture, so it now shows a `"` forcing the inline path against real length pressure instead of an artificial one.

🤖

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryArgumentsBuilderTests.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+        List<string> actual = await RetryArgumentsBuilder.BuildAttemptArgumentsAsync(
+            fileSystem.Object,
+            executableArguments,
+            indicesToCleanup,
+            "retry-root-2",
+            "retry-root",
+            "pipe-name",
+            ["failed uid", "failed-2"],
+            attemptCount: 2).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryArgumentsBuilderTests.cs:137]: Added. `BuildAttemptArgumentsAsync_WithNullFailedIds_KeepsOriginalFiltersAndMinimumExpectedTests` and the empty-array variant both take the first-attempt path and assert the whole argument list: the original `--filter-uid old-1 old-2` and `--treenode-filter /old-filter` survive, `--minimum-expected-tests 10` is not stripped, no extra `--filter-uid` is injected, and no response file is written.

🤖

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryOrchestratorHelperTests.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        ];
+
+        RetryOrchestratorHelper.RemoveOption(arguments, "target");
+
+        Assert.HasCount(3, arguments);
+        Assert.AreEqual("before", arguments[0]);
+        Assert.AreEqual("--other", arguments[1]);
+        Assert.AreEqual("after", arguments[2]);
+    }
+}
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryOrchestratorHelperTests.cs:99]: Done here rather than in a follow-up, since it is cheap.

`GetOptionArgumentIndex` gets four tests: long form, short form, short form winning when both are present, and no match when an option merely starts with the searched name. `FormatDelay` and `FormatDuration` get `[DataRow]` tests over the boundaries you named — 999 vs. 1000ms, whole seconds vs. leftover milliseconds, sub-second, sub-minute, and the minute format.

🤖

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryArgumentsBuilderTests.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+using Microsoft.Testing.Platform.Helpers;
+
+using Moq;
+
+namespace Microsoft.Testing.Extensions.UnitTests;
+
+[TestClass]
+public sealed class RetryArgumentsBuilderTests
+{
+    [DataRow(RetryCommandLineOptionsProvider.RetryFailedTestsMaxPercentageOptionName, "50")]
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryArgumentsBuilderTests.cs:23]: Added both, and a third one you did not ask for: `--retry-failed-tests-delay` as the last argument. That is the only place in `ComputeIndicesToCleanup` with an `argIndex + 1 < executableArguments.Length` guard, so it is the one branch that could return an index past the end of the array.

🤖

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryArgumentsBuilderTests.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            "3",
+            $"--{optionalOptionName}",
+            optionalOptionValue,
+            "--keep",
+            "value",
+        ];
+
+        List<int> actual = RetryArgumentsBuilder.ComputeIndicesToCleanup(executableArguments);
+
+        int[] expected = [1, 2, 3, 4];
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/UnitTests/Microsoft.Testing.Extensions.UnitTests/RetryArgumentsBuilderTests.cs:43]: Order is not part of the contract. The only consumer does `indexToCleanup.Contains(i)`, and the list is accumulated per option rather than sorted, so a returned order like `[9, 10, 1, 2]` is possible and asserting it would pin an implementation detail.

So I went with the equivalent-not-equal reading: `Assert.AreSequenceEqual(expected, actual, SequenceOrder.InAnyOrder)`. That is the MSTest equivalent of `CollectionAssert.AreEquivalent` — `CollectionAssert` itself is banned in this project by MSTEST0068. It also checks the count, so the separate `Assert.HasCount` is gone.

🤖

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

**Comment by nohwnd** [comment]
All five review threads are answered and resolved.

The one that mattered: the over-length fixture built a single 40,000-character argument, which is already past the 30,000 limit on its own, so the response-file test passed for the wrong reason and would have kept passing if the UID-length accounting were deleted. The failed-UID payload is now what crosses the limit, and I confirmed by removing that accumulation from RetryArgumentsBuilder that the test then fails.

The rest of the feedback is implemented rather than deferred: null and empty `lastListOfFailedId`, `ComputeIndicesToCleanup` with no and with all optional options, and `GetOptionArgumentIndex` / `FormatDelay` / `FormatDuration`.

This is my own PR so I cannot approve it. Ready for a second pair of eyes.

🤖

**Comment by nohwnd** [comment]
Both advisory reviews from this run are answered.

The parallel-safety audit found nothing to flag, so nothing to do there.

The test review is all A grades except two B's on `GetOptionArgumentIndex_WithLongForm_ReturnsOptionIndex` and `GetOptionArgumentIndex_WithShortForm_ReturnsOptionIndex`, where it suggests folding the two into one `[DataRow]`. I am keeping them separate. There are four `GetOptionArgumentIndex` tests and each one states a single rule of the contract, long form is found, short form is found, short form wins when both are present, and a prefix-only match is not a match. Folding two of the four leaves the group half parameterized and half named, and the test name is what tells you which rule broke when it fails. Both tests are 1/1 mutation killed, so the overlap is in shape only, it is not a coverage or a correctness point.

The second reason is cost, the required `microsoft.testfx` check is already three quarters of an hour into this commit with only the two slow Debug legs left, and pushing a rename plus one less method would discard that and start the whole run again. That is a bad trade for a test-only change that the same review calls "in strong shape".

Turning on auto-merge so this lands once the build finishes and it picks up an approval.

🤖

**Review by azat-msft — APPROVED** [review]
